### PR TITLE
Use tarball to transfer files to deploy

### PR DESCRIPTION
**Pull request by sebjulliand**
### Changes
Files to be deployed are put in a tarball (.tar archive) before being sent.
The benefits are:
- Faster deployment times, especially when deploying a lot of files
- Simpler deployment process (directory creations is handled by the `tar` command)
- File attributes and access permissions are kept during the transfer (if deploying from a POSIX os; not applicable for Windows)

Will fix https://github.com/halcyon-tech/vscode-ibmi/issues/1026

### Checklist
* [x] have tested my change
* [x] updated relevant documentation
* [x] **for feature PRs**: PR only includes one feature enhancement.

### CODE REVIEW

File: src/api/local/deployment.ts [hunk 1]
```diff
@@ -605,15 +510,46 @@ export namespace Deployment {
...
+      tar.create({ cwd: parameters.workspaceFolder.uri.fsPath, sync: true, file: localTarball.name }, toSend);
+      deploymentLog.appendLine(`Created deployment tarball ${localTarball.name}`);
+
+      progress?.report({ message: `sending deployment tarball...` });
+      await connection.client.putFile(localTarball.name, remoteTarball);
+      deploymentLog.appendLine(`Uploaded deployment tarball as ${remoteTarball}`);
+
+      progress?.report({ message: `extracting deployment tarball to ${parameters.remotePath}...` });
+      //Extract and remove tar's PaxHeader metadata folder
+      await connection.sendCommand({ command: `tar -xf ${remoteTarball} && rm -rf PaxHeader`, directory: parameters.remotePath });
```
Comment (worksofliam) [inline src/api/local/deployment.ts:539]: Which version of tar are we expecting to use here? More notes in final review.

**Review by worksofliam — CHANGES_REQUESTED** [review]
Before I approve, I think we need to be more explicit about which version of `tar` we should be using. There is the system shipped version, but also the version available in the IBM yum repo.

```
-bash-5.1$ /usr/bin/tar
Usage: tar -{c|r|t|u|x} [ -BdDEFhilmopRsUvwZ ] [ -Number ] [ -f TarFile ]
           [ -b Blocks ] [ -S [ Feet ] | [ Feet@Density ] | [ Blocksb ] ]
           [ -L InputList ] [-X ExcludeFile] [ -N Blocks ] [ -C Directory ] File ...
Usage: tar {c|r|t|u|x} [ bBdDEfFhilLXmNopRsSUvwZ[0-9] ]
           [ Blocks ] [ TarFile ] [ InputList ] [ ExcludeFile ] 
           [ [ Feet ] | [ Feet@Density ] | [ Blocksb ] ] [-C Directory ] File ...
-bash-5.1$ /QOpenSys/pkgs/bin/tar
/QOpenSys/pkgs/bin/tar: You must specify one of the '-Acdtrux', '--delete' or '--test-label' options
```

I think we need to update our internal API to ensure that `tar` is installed, which means adding `tar` to the `remoteApps` under `/usr/bin`. See `IBMi.ts:25`.

Thoughts?

**Review by worksofliam — COMMENTED** [review]
I've just done some functional testing and....... I am blown AWAY lol. This is so amazing.

Just to confirm, Seb - you are Windows right? I am mac and this is working very well.

File: src/api/local/deployment.ts [hunk 2]
```diff
@@ -605,15 +510,46 @@ export namespace Deployment {
...
+
+      const entries: string[] = [];
+      tar.t({ sync: true, file: localTarball.name, onentry: entry => entries.push(entry.path) });
+      deploymentLog.appendLine(`${entries.length} file(s) uploaded to ${parameters.remotePath}`);
+      entries.sort().map(e => `\t${e}`).forEach(deploymentLog.appendLine);
+    }
+    finally {
+      deploymentLog.appendLine('');
+      await connection.sendCommand({ command: `rm ${remoteTarball}` })
```
Comment (worksofliam) [inline src/api/local/deployment.ts:548]: I see you are deleting the remote file - this is great. Should we also delete the local `.tar` as well?

Comment (sebjulliand) [reply src/api/local/deployment.ts:548]: We do! This line below takes care of it:
`localTarball.removeCallback();`

File: src/api/local/deployment.ts (hunk 1 above)
Comment (sebjulliand) [reply src/api/local/deployment.ts:539]: That's the beauty of it: it doesn't mater.

**Comment by sebjulliand** [comment]
> Before I approve, I think we need to be more explicit about which version of `tar` we should be using. There is the system shipped version, but also the version available in the IBM yum repo.
> 
> ```
> -bash-5.1$ /usr/bin/tar
> Usage: tar -{c|r|t|u|x} [ -BdDEFhilmopRsUvwZ ] [ -Number ] [ -f TarFile ]
>            [ -b Blocks ] [ -S [ Feet ] | [ Feet@Density ] | [ Blocksb ] ]
>            [ -L InputList ] [-X ExcludeFile] [ -N Blocks ] [ -C Directory ] File ...
> Usage: tar {c|r|t|u|x} [ bBdDEfFhilLXmNopRsSUvwZ[0-9] ]
>            [ Blocks ] [ TarFile ] [ InputList ] [ ExcludeFile ] 
>            [ [ Feet ] | [ Feet@Density ] | [ Blocksb ] ] [-C Directory ] File ...
> -bash-5.1$ /QOpenSys/pkgs/bin/tar
> /QOpenSys/pkgs/bin/tar: You must specify one of the '-Acdtrux', '--delete' or '--test-label' options
> ```
> 
> I think we need to update our internal API to ensure that `tar` is installed, which means adding `tar` to the `remoteApps` under `/usr/bin`. See `IBMi.ts:25`.
> 
> Thoughts?

I doesn't mater which one is found and used. Both are supporting the POSIX flags `x` and `f`. And I think tar is always available out of the box. But it wouldn't hurt adding to the `remoteApps` 😊

**Comment by sebjulliand** [comment]
> I've just done some functional testing and....... I am blown AWAY lol. This is so amazing.
> 
> Just to confirm, Seb - you are Windows right? I am mac and this is working very well.

Thanks! Yes, I'm on Windows and I tested it on Linux too; it works very well in both cases.
Thankfully, the JS `tar` implementation is os independent.

**Comment by sebjulliand** [comment]
@worksofliam I added tar to the APi's `remoteFeatures`, it will look in both possible locations.
I tested both just to be sure and they work the same, as expected. I'll let you give it a try!

**Review by worksofliam — APPROVED** [review]
I am happy with my mac testing. Tried every deploy option and it's all working smoothly. If you are happy that it works on Windows Seb, then let's merge.

Thanks for your great work here!

p.s. I tried with dumbo_app and the improvement is major. You should compare for yourself. You might get a chuckle from it.

**Pull request merged**